p2p/protocols/protocol.go

```diff
@@ -217,39 +225,127 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
 		decode = decodeWithoutContext
 	}
 	return &Peer{
-		Peer:   peer,
-		rw:     rw,
-		spec:   spec,
-		encode: encode,
-		decode: decode,
+		Peer:     peer,
+		rw:       rw,
+		spec:     spec,
+		encode:   encode,
+		decode:   decode,
+		shutdown: make(chan struct{}),
 	}
 }
 
-// Run starts the forever loop that handles incoming messages
-// called within the p2p.Protocol#Run function
-// the handler argument is a function which is called for each message received
+// Run starts the forever loop that handles incoming messages.
+// The handler argument is a function which is called for each message received
 // from the remote peer, a returned error causes the loop to exit
-// resulting in disconnection
+// resulting in disconnection of the protocol
+// there can be only one run loop per peer.
+// It returns the error that caused the loop to exit.
 func (p *Peer) Run(handler func(ctx context.Context, msg interface{}) error) error {
-	for {
-		if err := p.handleIncoming(handler); err != nil {
-			if err != io.EOF {
-				metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
-				log.Error("peer.handleIncoming", "err", err)
+	p.setRunning(true)
+
+	done := make(chan error, 1)
+	go func() {
```

Comment:
Nice idea, I will try something like that. It's not that simple, again because of a bit fuzzy shutdown. ReadMsg function is a blocking one, and we need to listen if either this one, or one of the async routines return an error and quit. This is why I added shutdown channel as well. On the other hand, we don't want to dispatch any more async jobs if shutdown was called, and this iis why there is also a running variable. Basically, the current situation on the shutdown, is that we want loop to exit with nil, but disable dispatching async jobs in the moment we are waiting for wg. Or maybe I am missing somethinig 🙈  I will play around this idea a bit and try to make it work!

Reply:
I have reduced the logic to one global channel, for both errc and shutdown, which makes the code look like this now :) If we need to separate them again when rewriting shutdown logic, we can do it similarly like this as well, with just one  select statement  instead of returning just errc. Thanks for the tip!